File: src/ee/core/Platform.cpp
//
//  PlatformUtils.cpp
//  Pods
//
//  Created by eps on 6/16/20.
//

#include "ee/core/Platform.hpp"

#include <ee/nlohmann/json.hpp>

#include "ee/core/IMessageBridge.hpp"
#include "ee/core/Task.hpp"
#include "ee/core/Utils.hpp"

namespace ee {
namespace core {
namespace {
// clang-format off
const std::string kPrefix = "Platform_";

const auto kIsApplicationInstalled   = kPrefix + "isApplicationInstalled";
const auto kOpenApplication          = kPrefix + "openApplication";

const auto kGetApplicationId         = kPrefix + "getApplicationId";
const auto kGetApplicationName       = kPrefix + "getApplicationName";
const auto kGetVersionName           = kPrefix + "getVersionName";
const auto kGetVersionCode           = kPrefix + "getVersionCode";

const auto kGetApplicationSignatures = kPrefix + "getApplicationSignatures";
const auto kIsInstantApp             = kPrefix + "isInstantApp";
const auto kIsTablet                 = kPrefix + "isTablet";
const auto kGetDensity               = kPrefix + "getDensity";
const auto kGetDeviceId              = kPrefix + "getDeviceId";
const auto kGetSafeInset             = kPrefix + "getSafeInset";

const auto kSendMail                 = kPrefix + "sendMail";
const auto kTestConnection           = kPrefix + "testConnection";
const auto kShowInstallPrompt        = kPrefix + "showInstallPrompt";
// clang-format on
} // namespace

using Self = Platform;

IMessageBridge* Self::bridge_ = nullptr;

void Self::registerHandlers(IMessageBridge& bridge) {
    bridge_ = &bridge;
}

bool Self::isApplicationInstalled(const std::string& applicationId) {
    auto response = bridge_->call(kIsApplicationInstalled, applicationId);
    return toBool(response);
}

bool Self::openApplication(const std::string& applicationId) {
    auto response = bridge_->call(kOpenApplication, applicationId);
    return toBool(response);
}

std::string Self::getApplicationId() {
    return bridge_->call(kGetApplicationId);
}

std::string Self::getApplicationName() {
    return bridge_->call(kGetApplicationName);
}

std::string Self::getVersionName() {
    return bridge_->call(kGetVersionName);
}

std::string Self::getVersionCode() {
    return bridge_->call(kGetVersionCode);
}

std::string Self::getSHA1CertificateFingerprint() {
#ifdef EE_X_ANDROID
    nlohmann::json request;
    request["algorithm"] = "SHA1";
    auto response = bridge_->call(kGetApplicationSignatures, request.dump());
    auto json = nlohmann::json::parse(response);
    auto signatures = json["signatures"];
    return signatures.empty() ? "" : signatures[0];
#else  // EE_X_ANDROID
    return "";
#endif // EE_X_ANDROID
}

bool Self::isInstantApp() {
#ifdef EE_X_ANDROID
    auto response = bridge_->call(kIsInstantApp);
    return toBool(response);
#else  // EE_X_ANDROID
    return false;
#endif // EE_X_ANDROID
}

bool Self::isTablet() {
    auto response = bridge_->call(kIsTablet);
    return toBool(response);
}

float Self::getDensity() {
    auto response = bridge_->call(kGetDensity);
    return std::stof(response);
}

Task<std::string> Self::getDeviceId() {
    auto response = co_await bridge_->callAsync(kGetDeviceId);
    co_return response;
}

SafeInset Self::getSafeInset() {
    SafeInset result;
#ifdef EE_X_ANDROID
    auto response = bridge_->call(kGetSafeInset);
    auto json = nlohmann::json::parse(response);

    result.left = json["left"];
    result.right = json["right"];
    result.top = json["top"];
    result.bottom = json["bottom"];
#else  // EE_X_ANDROID
    // TODO.
    result.left = 0;
    result.right = 0;
    result.top = 0;
    result.bottom = 0;
#endif // EE_X_ANDROID
    return result;
}

bool Self::sendMail(const std::string& recipient, const std::string& subject,
                    const std::string& body) {
    nlohmann::json json;
    json["recipient"] = recipient;
    json["subject"] = subject;
    json["body"] = body;
    auto response = bridge_->call(kSendMail, json.dump());
    return toBool(response);
}

Task<bool> Self::testConnection(const std::string& hostName, float timeOut) {
    nlohmann::json json;
    json["host_name"] = hostName;
    json["time_out"] = timeOut;
    auto response = co_await bridge_->callAsync(kTestConnection, json.dump());
    co_return toBool(response);
}

void Self::showInstallPrompt(const std::string& url,
                             const std::string& referrer) {
#ifdef EE_X_ANDROID
    nlohmann::json json;
    json["url"] = url;
    json["referrer"] = referrer;
    bridge_->call(kShowInstallPrompt, json.dump());
#endif // EE_X_ANDROID
}
} // namespace core
} // namespace ee

File: src/ios/ee/core/Platform.swift
//
//  PlatformUtils.swift
//  Pods
//
//  Created by eps on 6/16/20.
//

import Foundation
import Reachability
import RxSwift

public class Platform: NSObject {
    private static let kPrefix = "Platform_"

    private static let kIsApplicationInstalled = kPrefix + "isApplicationInstalled"
    private static let kOpenApplication = kPrefix + "openApplication"
    private static let kGetApplicationId = kPrefix + "getApplicationId"
    private static let kGetApplicationName = kPrefix + "getApplicationName"
    private static let kGetVersionName = kPrefix + "getVersionName"
    private static let kGetVersionCode = kPrefix + "getVersionCode"
    private static let kIsTablet = kPrefix + "isTablet"
    private static let kGetDensity = kPrefix + "getDensity"
    private static let kGetViewSize = kPrefix + "getViewSize"
    private static let kGetScreenSize = kPrefix + "getScreenSize"
    private static let kGetDeviceId = kPrefix + "getDeviceId"
    private static let kSendMail = kPrefix + "sendMail"
    private static let kTestConnection = kPrefix + "testConnection"

    public class func registerHandlers(_ bridge: IMessageBridge) {
        bridge.registerHandler(kIsApplicationInstalled) { message in
            let applicationId = message
            return Utils.toString(isApplicationInstalled(applicationId))
        }
        bridge.registerHandler(kOpenApplication) { message in
            let applicationId = message
            return Utils.toString(openApplication(applicationId))
        }
        bridge.registerHandler(kGetApplicationId) { _ in getApplicationId() }
        bridge.registerHandler(kGetApplicationName) { _ in getApplicationName() }
        bridge.registerHandler(kGetVersionName) { _ in getVersionName() }
        bridge.registerHandler(kGetVersionCode) { _ in getVersionCode() }
        bridge.registerHandler(kIsTablet) { _ in Utils.toString(isTablet()) }
        bridge.registerHandler(kGetDensity) { _ in "\(getDensity())" }
        bridge.registerHandler(kGetViewSize) { _ in
            let size = self.getViewSize()
            return EEJsonUtils.convertDictionary(toString: [
                "width": size.width,
                "height": size.height
            ])
        }
        bridge.registerHandler(kGetScreenSize) { _ in
            let size = self.getScreenSize()
            return EEJsonUtils.convertDictionary(toString: [
                "width": size.width,
                "height": size.height
            ])
        }
        bridge.registerAsyncHandler(kGetDeviceId) { _, resolver in
            resolver(getDeviceId())
        }
        bridge.registerHandler(kSendMail) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard
                let recipient = dict["recipient"] as? String,
                let subject = dict["subject"] as? String,
                let body = dict["body"] as? String
            else {
                assert(false, "Unexpected value")
                return ""
            }
            return Utils.toString(sendMail(recipient, subject, body))
        }
        bridge.registerAsyncHandler(kTestConnection) { message, resolver in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard
                let hostName = dict["host_name"] as? String,
                let timeOut = dict["time_out"] as? Float
            else {
                assert(false, "Unexpected value")
                return
            }
            testConnection(hostName, timeOut)
                .subscribe(
                    onSuccess: {
                        result in resolver(Utils.toString(result))
                    }, onError: {
                        _ in resolver(Utils.toString(false))
                    })
        }
    }

    private class func isApplicationInstalled(_ applicationId: String) -> Bool {
        #if os(iOS)
        if let url = URL(string: applicationId + "://") {
            return UIApplication.shared.canOpenURL(url)
        }
        #else
        if NSWorkspace.shared.fullPath(forApplication: applicationId) != nil {
            return true
        }
        #endif
        return false
    }

    private class func openApplication(_ applicationId: String) -> Bool {
        if let url = URL(string: applicationId + "://") {
            #if os(iOS)
            return UIApplication.shared.openURL(url)
            #elseif os(OSX)
            return NSWorkspace.shared.open(url)
            #endif
        }
        return false
    }

    private class func getApplicationId() -> String {
        let key = kCFBundleIdentifierKey as String
        return Bundle.main.object(forInfoDictionaryKey: key) as? String ?? ""
    }

    private class func getApplicationName() -> String {
        let key = kCFBundleNameKey as String
        return Bundle.main.object(forInfoDictionaryKey: key) as? String ?? ""
    }

    private class func getVersionName() -> String {
        let key = "CFBundleShortVersionString"
        return Bundle.main.object(forInfoDictionaryKey: key) as? String ?? ""
    }

    private class func getVersionCode() -> String {
        let key = kCFBundleVersionKey as String
        return Bundle.main.object(forInfoDictionaryKey: key) as? String ?? ""
    }

    private class func isTablet() -> Bool {
        #if os(iOS)
        return UIDevice.current.model.hasPrefix("iPad")
        #else
        return false
        #endif
    }

    @objc
    public class func getDensity() -> Float {
        #if os(iOS)
        return Float(UIScreen.main.scale)
        #else
        let screen = NSScreen.screens.max { lhs, rhs in
            lhs.backingScaleFactor < rhs.backingScaleFactor
        }
        return Float(screen?.backingScaleFactor ?? 1)
        #endif
    }

    public class func getViewSize() -> CGSize {
        return getScreenSize()
    }

    public class func getScreenSize() -> CGSize {
        let scale = getDensity()
        let size = UIScreen.main.bounds.size
        return CGSize(width: size.width * CGFloat(scale), height: size.height * CGFloat(scale))
    }

    private class func getDeviceId() -> String {
        #if os(iOS)
        return UIDevice.current.identifierForVendor?.uuidString ?? ""
        #else
        // Get the platform expert
        let platformExpert: io_service_t = IOServiceGetMatchingService(kIOMasterPortDefault, IOServiceMatching("IOPlatformExpertDevice"))

        // Get the serial number as a CFString ( actually as Unmanaged<AnyObject>! )
        let serialNumberAsCFString = IORegistryEntryCreateCFProperty(
            platformExpert, kIOPlatformSerialNumberKey as CFString, kCFAllocatorDefault, 0)

        // Release the platform expert (we're responsible)
        IOObjectRelease(platformExpert)

        // Take the unretained value of the unmanaged-any-object
        // (so we're not responsible for releasing it)
        // and pass it back as a String or, if it fails, an empty string
        return (serialNumberAsCFString?.takeUnretainedValue() as? String) ?? ""
        #endif
    }

    private class func sendMail(_ recipient: String, _ subject: String, _ body: String) -> Bool {
        let str = String("mailto:\(recipient)?subject=\(subject)&body=\(body)")
        guard let encodedStr = str.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else {
            return false
        }
        guard let url = URL(string: encodedStr) else {
            return false
        }
        #if os(iOS)
        return UIApplication.shared.openURL(url)
        #else
        return NSWorkspace.shared.open(url)
        #endif
    }

    private class func testConnection(_ hostName: String, _ timeOut: Float) -> Single<Bool> {
        let scheduler = SerialDispatchQueueScheduler(qos: .background)
        return Single<Bool>.create { single in
            do {
                let status = try Reachability(hostname: hostName).connection
                let isReachable = status != .unavailable
                single(.success(isReachable))
            } catch let ex {
                single(.error(ex))
            }
            return Disposables.create()
        }
        .subscribeOn(scheduler)
        .timeout(.milliseconds(Int(timeOut * 1000)), scheduler: scheduler)
    }
}
